src/fvom_init.F90

```diff
@@ -872,150 +918,269 @@ subroutine find_levels_cavity(mesh)
         
         !_______________________________________________________________________
         ! vertical elem level index of cavity-ocean boundary
-        exit_flag=0
         ulevels(elem) = 1
+        if (dmean<0.0_WP) ulevels(elem) = 2
+        
         do nz=1,nlevels(elem)-1
-            !!PS if(Z(nz)<dmean) then
-            if(Z(nz)<dmean .or. nlevels(elem)-nz<=3) then
-                exit_flag=1
-                ulevels(elem)=nz
+            if (Z(nz)<dmean .or. nlevels(elem)-nz<=3) then
+                ulevels(elem)=nz         !    to compute shechpetkin PGF
                 exit
             end if
         end do
-        if ((exit_flag==0).and.(dmean<0)) ulevels(elem)=nlevels(elem)     
         cavity_maxlev = max(cavity_maxlev,ulevels(elem))
     end do
     
+    !___________________________________________________________________________
+    ! write out cavity mesh files for vertice and elemental position of 
+    ! vertical cavity-ocean boundary before the iterative geometric adaption to 
+    ! eliminate isolated cells
+    if (mype==0) then
+        ! write out elemental cavity-ocean boundary level
+        file_name=trim(meshpath)//'cavity_elvls_raw.out'
+        open(20, file=file_name)
+        do elem=1,elem2D
+            write(20,*) ulevels(elem)
+        enddo
+        close(20)
+    endif
+    
     !___________________________________________________________________________
     ! Eliminate cells that have two cavity boundary faces --> should not be 
     ! possible in FESOM2.0
     ! loop over all cavity levels
-    do nz=1,cavity_maxlev
-        exit_flag=0
-        count_iter=0
+    allocate(elemreducelvl(elem2d),elemfixlvl(elem2d))
+    allocate(numelemtonode(nl,nod2D),idxelemtonode(nl,nod2D))
```

Comment:
i can convert elemreducelvl(elem2d) and elemfixlvl(elem2d) to logical array should save some memory

and 

I can convert the code so that i only need numelemtonode(nl),idxelemtonode(nl)) which should save a lot. But anyway the model will only step into this part when cavity is used